packages/fxa-auth-server/lib/payments/stripe.js

```diff
@@ -1015,6 +1011,179 @@ class StripeHelper {
+    const customer = await this.expandResource(
+      subscription.customer,
+      'customers'
+    );
+    if (customer.deleted === true) {
+      throw error.unknownCustomer(subscription.customer);
+    }
+
+    const {
+      email,
+      metadata: { userid: uid },
+    } = customer;
+
+    const previousAttributes = eventData.previous_attributes;
+    const planNew = subscription.plan;
+    const planOld = previousAttributes.plan;
+    const cancelAtPeriodEndNew = subscription.cancel_at_period_end;
+    const cancelAtPeriodEndOld = previousAttributes.cancel_at_period_end;
+
+    const abbrevProductNew = await this.expandAbbrevProductForPlan(planNew);
+    const { interval: productPaymentCycleNew, amount: planAmountNew } = planNew;
+    const {
+      product_id: productIdNew,
+      product_name: productNameNew,
+    } = abbrevProductNew;
+    const {
+      productOrder: productOrderNew,
+      emailIconURL: productIconURLNew = '',
+      downloadURL: productDownloadURLNew = '',
+    } = this.mergeMetadata(planNew, abbrevProductNew);
+
+    const baseDetails = {
+      uid,
+      email,
+      productIdNew,
+      productNameNew,
+      productIconURLNew,
+      productDownloadURLNew,
+      productPaymentCycleNew,
+      planAmountNew,
+      closeDate: event.created,
+    };
+
+    if (!cancelAtPeriodEndOld && cancelAtPeriodEndNew && !planOld) {
+      return this.extractSubscriptionUpdateCancellationDetailsForEmail(
+        subscription,
+        baseDetails
+      );
+    } else if (cancelAtPeriodEndOld && !cancelAtPeriodEndNew && !planOld) {
+      return this.extractSubscriptionUpdateReactivationDetailsForEmail(
+        subscription,
+        baseDetails
+      );
+    } else if (!cancelAtPeriodEndOld && !cancelAtPeriodEndNew && planOld) {
+      return this.extractSubscriptionUpdateUpgradeDowngradeDetailsForEmail(
+        subscription,
+        baseDetails,
+        customer,
+        productOrderNew,
+        planOld
+      );
+    }
+
+    // unknown update scenario, but let's return some details anyway
+    return baseDetails;
+  }
+
+  async extractSubscriptionUpdateCancellationDetailsForEmail(
+    subscription,
+    baseDetails
+  ) {
+    // https://github.com/mozilla/subhub/blob/e224feddcdcbafaf0f3cd7d52691d29d94157de5/src/hub/vendor/customer.py#L598
+    return {
+      ...baseDetails,
+      isCancellation: true,
+      cancelledAt: subscription.canceled_at,
+      cancelAt: subscription.cancel_at,
+      cancelAtPeriodEnd: subscription.cancel_at_period_end,
+      currentPeriodStart: subscription.current_period_start,
+      currentPeriodEnd: subscription.current_period_end,
+      invoiceId: subscription.latest_invoice,
+    };
+  }
+
+  async extractSubscriptionUpdateReactivationDetailsForEmail(
+    subscription,
+    baseDetails
+  ) {
+    // https://github.com/mozilla/subhub/blob/e224feddcdcbafaf0f3cd7d52691d29d94157de5/src/hub/vendor/customer.py#L625
+    const invoice = await this.expandResource(
```

Comment:
I think it would be better here to not use `expandResource` and instead just do something like this, which would mean that stripe would be called at most once:

```suggestion
    let invoice = subscription.latest_invoice;
    if (typeof invoice === 'string') {
       // if we have to do a fetch, go ahead and ensure we also get the additional needed resource
        const invoice = await this.stripe.invoices.retrieve(invoice, { expand: ['charge'] });
    }
    
    // this will only result to a call to stripe if the invoice was already expanded, but the charge was not
    const charge = await this.expandResource(invoice.charge, CHARGES_RESOURCE);
```

Reply:
I should probably just always fetch here, since the data from webhook events seems to never have an expanded `latest_invoice`

Reply:
Yeah, as far as I know it's not, but of course, nothing in Stripe confirms/denies this 😂